numpy/lib/function_base.py

```diff
@@ -1574,17 +1609,38 @@ def gradient(f, *varargs, **kwargs):
 
     n = len(varargs)
     if n == 0:
-        dx = [1.0]*N
+        dx = [1.0] * N
     elif n == 1:
-        dx = [varargs[0]]*N
+        if np.isscalar(varargs[0]):
+            dx = [varargs[0]] * N
+        else:
+            if(len(varargs[0]) != f.shape[0]) or len(axes) != 1:
+                raise ValueError("distances must be either scalars or match "
+                                 "the length of the corresponding dimension")
+            diffx = np.diff(varargs[0])
+            # if distances are constant reduce to the scalar case
```

Comment:
I have done a couple of [tests](https://gist.github.com/apbard/6323e7390fa98033e74230e1db86ed82) and it actually seems to provide a non-negligible speed-up:

    np.gradient 0.835474014282
    gradient_opt scalar 0.849312067032
    gradient_noopt scalar 0.865785121918
    gradient_opt vector 0.984735012054 
    gradient_noopt vector 1.54027509689

 I think that it is probably worth keeping it.